File: Matt Zanchelli/MTZTimelineEventView.m
//
//  MTZTimelineEventView.m
//  Matt Zanchelli
//
//  Created by Matt Zanchelli on 4/8/14.
//  Copyright (c) 2014 Matt Zanchelli. All rights reserved.
//

#import "MTZTimelineEventView.h"

#import "PSPushPopPressView/PSPushPopPressView.h"

@interface MTZTimelineEventView ()

///	Label for the date.
@property (strong, nonatomic) UILabel *dateLabel;

///	Label for the title.
@property (strong, nonatomic) UILabel *titleLabel;

///	Label for the description.
@property (strong, nonatomic) UILabel *descriptionLabel;

///	The view to load content to expand/manipulate.
@property (strong, nonatomic) PSPushPopPressView *pushPopPressView;

///	The image view to put the image in.
@property (strong, nonatomic) UIImageView *imageView;

///	The mark on the timeline for the event.
@property (strong, nonatomic) UIImageView *dotView;

///	The vertical stripe representing the timeline.
@property (strong, nonatomic) UIView *timelineStripeView;

@end

@implementation MTZTimelineEventView

- (id)initWithTimelineEvent:(MTZTimelineEvent *)timelineEvent
{
	self = [super initWithFrame:CGRectMake(0, 0, 1024, 130)];
	if (self) {
		[self setUpMTZTimeLineEventView];
		
		self.date = timelineEvent.date;
		self.title = timelineEvent.title;
		self.description = timelineEvent.description;
		self.image = timelineEvent.image;
	}
	return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
	self = [super initWithCoder:aDecoder];
    if (self) {
        // Initialization code
		[self setUpMTZTimeLineEventView];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
		[self setUpMTZTimeLineEventView];
    }
    return self;
}

- (id)init
{
	self = [super init];
	if (self) {
        // Initialization code
		[self setUpMTZTimeLineEventView];
    }
    return self;
}

- (void)setUpMTZTimeLineEventView
{	
	// Set up frame.
	self.frame = CGRectMake(self.frame.origin.x, self.frame.origin.y, 1024, 130);
	
	self.timelineStripeView = [[UIView alloc] initWithFrame:CGRectMake(128, 0, 1, 130)];
	[self addSubview:self.timelineStripeView];
	
	// Add dot view.
	self.dotView = [[UIImageView alloc] initWithFrame:CGRectMake(123, 11, 11, 11)];
	self.dotView.image = [UIImage imageNamed:@"Timeline-Dot"];
	[self addSubview:self.dotView];
	
	// Add PPP view.
	self.pushPopPressView = [[PSPushPopPressView alloc] initWithFrame:CGRectMake(660, 0, 344, 150)];
	self.pushPopPressView.backgroundColor = [UIColor whiteColor];
	[self addSubview:self.pushPopPressView];
	
	// Add image view.
	self.imageView = [[UIImageView alloc] initWithFrame:self.pushPopPressView.bounds];
	self.imageView.backgroundColor = [UIColor whiteColor];
	self.imageView.opaque = YES;
	self.imageView.clipsToBounds = YES;
	self.imageView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
	self.imageView.contentMode = UIViewContentModeScaleAspectFit;// UIViewContentModeScaleAspectFill;
#warning The delegate should be the corresponding view controller.
//	self.imageView.pushPopPressViewDelegate = self;
	[self.pushPopPressView addSubview:self.imageView];
	
	// Set up labels frames.
	self.dateLabel = [[UILabel alloc] initWithFrame:CGRectMake(21, 5, 88, 25)];
	self.titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(148, 5, 472, 25)];
	self.descriptionLabel = [[UILabel alloc] initWithFrame:CGRectMake(148, 29, 472, 100)];
	
	// Set backgrounds.
	self.dateLabel.backgroundColor = [UIColor whiteColor];
	self.titleLabel.backgroundColor = [UIColor whiteColor];
	self.descriptionLabel.backgroundColor = [UIColor whiteColor];
	self.dotView.backgroundColor = [UIColor whiteColor];
	self.timelineStripeView.backgroundColor = [UIColor colorWithWhite:0.85f alpha:1.0f];
	
	// Make views opaque.
	self.dateLabel.opaque = YES;
	self.titleLabel.opaque = YES;
	self.descriptionLabel.opaque = YES;
	self.dotView.opaque = YES;
	self.timelineStripeView.opaque = YES;
	
	// Set up number of lines.
	self.dateLabel.numberOfLines = 1;
	self.titleLabel.numberOfLines = 1;
	self.descriptionLabel.numberOfLines = 0;
	
	// Add to view hierachy.
	[self addSubview:self.dateLabel];
	[self addSubview:self.titleLabel];
	[self addSubview:self.descriptionLabel];
	
	// Set default text.
	[self setDate:@"Feb 1993"];
	[self setTitle:@"Event Title."];
	[self setDescription:@"Event description."];
}


#pragma mark - Properties

- (void)setDate:(NSString *)date
{
	_date = date;
	self.dateLabel.text = date;
	[self.dateLabel applyMTZStyle:MTZStyleEventDate];
}

- (void)setTitle:(NSString *)title
{
	_title = title;
	self.titleLabel.text = title;
	[self.titleLabel applyMTZStyle:MTZStyleEventTitle];
}

- (void)setDescription:(NSString *)description
{
	_description = description;
	self.descriptionLabel.text = description;
	[self.descriptionLabel applyMTZStyle:MTZStyleEventDetail];
}

- (void)setImage:(UIImage *)image
{
	_image = image;
#warning set the size of imageView to fit properly.
	self.imageView.image = image;
}


@end

File: Matt Zanchelli/MTZTimelineEventView.swift
//
//  MTZTimelineEventView.swift
//  Matt Zanchelli
//
//  Created by Matt Zanchelli on 6/5/14.
//  Copyright (c) 2014 Matt Zanchelli. All rights reserved.
//

import UIKit

protocol MTZTimelineEventViewInterface {
	///	Set up a timeline event view with a timeline event object.
	///	@param timelineEvent The timeline event object to create the view from.
	func setUpWithTimelineEvent(timelineEvent: MTZTimelineEvent)
	
	///	Date of the event (as a string).
	var date : NSString { get set }
	
	///	The title of the event.
	var title : NSString { get set }
	
	///	Description of the event.
	var descriptionString : NSString { get set }
	
	///	An image corresponding to the event.
	var image : UIImage? { get set }
}

class MTZTimelineEventView: UIView, MTZTimelineEventViewInterface {
	
	// MARK: Public
	
	var date : NSString = "Feb 1993" {
	didSet {
		dateLabel.text = date
		dateLabel.applyMTZStyleEventDate()
	}
	}
	
	var title : NSString = "Event Title." {
	didSet {
		titleLabel.text = title
		titleLabel.applyMTZStyleEventTitle()
	}
	}
	
	var descriptionString : NSString = "Event description." {
	didSet {
		descriptionLabel.text = descriptionString
		descriptionLabel.applyMTZStyleEventDetail()
	}
	}
	
	var image : UIImage? {
	didSet {
		if let img = image {
			photoView.image = img
			let imageRect = CGRect(x: 0, y: 0, width: img.size.width, height: img.size.height)
			pushPopPressView.frame = CGRectScaledRectToFitInRect(imageRect, pushPopPressViewFrame)
		} else {
			photoView.image = nil
		}
	}
	}
	
	func setUpWithTimelineEvent(timelineEvent: MTZTimelineEvent) {
		date = timelineEvent.date;
		title = timelineEvent.title;
		descriptionString = timelineEvent.description;
		image = timelineEvent.image;
	}
	
	
	// MARK: Initializers
	
	init(frame: CGRect) {
		var modifiedFrame = frame
		modifiedFrame.size.width = 1024
		modifiedFrame.size.height = 130
		super.init(frame: modifiedFrame)
		setup()
	}
	
	func setup() {
		addSubview(timelineStripeView)
		addSubview(dotView)
		addSubview(pushPopPressView)
		pushPopPressView.addSubview(photoView)
		addSubview(dateLabel)
		addSubview(titleLabel)
		addSubview(descriptionLabel)
	}
	
	
	// MARK: Private
	
	///	Label for the date.
	let dateLabel : UILabel = {
		let view = UILabel(frame:CGRect(x: 21, y: 5, width: 88, height: 25))
		view.backgroundColor = UIColor.whiteColor()
		view.opaque = true
		view.numberOfLines = 1
		return view
	}()
	
	///	Label for the title.
	let titleLabel : UILabel = {
		let view = UILabel(frame:CGRect(x: 148, y: 5, width: 472, height: 25))
		view.backgroundColor = UIColor.whiteColor()
		view.opaque = true
		view.numberOfLines = 1
		return view
	}()
	
	///	Label for the description.
	let descriptionLabel : UILabel = {
		let view = UILabel(frame:CGRect(x: 148, y: 29, width: 472, height: 100))
		view.backgroundColor = UIColor.whiteColor()
		view.opaque = true
		view.numberOfLines = 0
		return view
	}()
	
	///	The view to load content to expand/manipulate.
	let pushPopPressView : PSPushPopPressView = {
//		let view = PSPushPopPressView(frame:pushPopPressViewFrame)
		let view = PSPushPopPressView(frame:CGRect(x: 660, y: 0, width: 344, height: 150))
		view.delegate = UIApplication.sharedApplication().keyWindow.rootViewController as PSPushPopPressViewDelegate
		view.backgroundColor = UIColor.whiteColor()
		view.retainAspectRatio = true
		return view
	}()
	
	///	The original frame for @c pushPopPressView.
	let pushPopPressViewFrame : CGRect = CGRect(x: 660, y: 0, width: 344, height: 150)
	
	///	The image view to put the image in.
	let photoView : UIImageView = {
//		let view = UIImageView(frame:pushPopPressView.bounds)
		let view = UIImageView(frame:CGRect(x: 0, y: 0, width: 344, height: 150))
		view.backgroundColor = UIColor.whiteColor()
		view.opaque = true
		view.clipsToBounds = true
		view.autoresizingMask = .FlexibleWidth | .FlexibleHeight;
		view.contentMode = .ScaleAspectFit
		return view
	}()
	
	///	The mark on the timeline for the event.
	let dotView : UIImageView = {
		let view = UIImageView(frame:CGRect(x: 123, y: 11, width: 11, height: 11))
		view.image = UIImage(named:"Timeline-Dot")
		view.backgroundColor = UIColor.whiteColor()
		view.opaque = true
		return view
	}()
	
	///	The vertical stripe representing the timeline.
	let timelineStripeView : UIView = {
		let view = UIView(frame:CGRect(x: 128, y: 0, width: 1, height: 130))
		view.backgroundColor = UIColor(white:0.85, alpha:1)
		view.opaque = true
		return view
	}()
	
}
